src/main/java/com/mongodb/spark/sql/connector/read/MongoStreamPartitionReader.java

```diff
@@ -0,0 +1,204 @@
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+
+    this.readConfig = readConfig;
+    this.currentRow = null;
+    this.lastOffset =
+        new ResumeTokenPartitionOffset(partition.getResumeTokenOffset().getResumeToken());
+
+    LOGGER.debug(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  @Override
+  public PartitionOffset getOffset() {
+    Assertions.ensureState(
+        () -> !closed, () -> "Cannot call getOffset() on a closed PartitionReader.");
+    LOGGER.trace("getOffset called, returning: {}", lastOffset);
+    return lastOffset;
+  }
+
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    return withCursor(
+        c -> {
+          try {
+            BsonDocument next = c.next();
+            lastOffset = new ResumeTokenPartitionOffset(c.getResumeToken());
+            if (next.getString("operationType", new BsonString(""))
+                .getValue()
+                .toLowerCase(Locale.ROOT)
+                .equals(INVALIDATE)) {
+              LOGGER.info(
+                  "Change stream cursor has been invalidated. This happens when a collection is dropped. Closing cursor.");
+              lastOffset = EMPTY_RESUME_TOKEN_PARTITION_OFFSET;
+              releaseCursor();
+              releaseClient();
+            }
+            currentRow = bsonDocumentToRowConverter.toInternalRow(next);
+          } catch (NoSuchElementException e) {
+            LOGGER.warn("No such element! {}", e.getMessage());
+            return false;
+          }
+          return true;
+        });
+  }
+
+  @Override
+  public InternalRow get() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call get() on a closed PartitionReader.");
+    Assertions.ensureState(
+        () -> currentRow != null,
+        () -> "Current row is null, this should not happen if `next()` returns false.");
+    LOGGER.trace("Get called: {}", currentRow);
+    return currentRow;
+  }
+
+  @Override
+  public void close() {
+    if (!closed) {
+      closed = true;
+      releaseCursor();
+      releaseClient();
+    }
+  }
+
+  private MongoChangeStreamCursor<BsonDocument> getCursor() {
+    if (mongoClient == null) {
+      mongoClient = readConfig.getMongoClient();
+    }
+    if (changeStreamCursor == null) {
+      LOGGER.info("Opened change stream cursor for partitionId: {}", partition.getPartitionId());
+      ChangeStreamIterable<Document> changeStreamIterable =
+          mongoClient
+              .getDatabase(readConfig.getDatabaseName())
+              .getCollection(readConfig.getCollectionName())
+              .watch(partition.getPipeline());
+
+      if (!lastOffset.getResumeToken().isEmpty()) {
+        changeStreamIterable = changeStreamIterable.resumeAfter(lastOffset.getResumeToken());
+      }
```

Comment:
Is my understanding correct, that when the offset is empty and no new writes are happening, `changeStreamIterable` will not return any documents, and, therefore, `MongoStreamPartitionReader` won't read any data?

- If yes, then how can this reader read any data, given that the offset comes from the Spark framework via the `MongoContinuousStream.planInputPartitions` method, and Spark framework cannot pass any offset other than `MongoContinuousStream.initialOffset` unless it got the offset previously from `MongoContinuousStream.mergeOffsets` (merged offsets come from `MongoStreamPartitionReader.getOffset`)?
- If no, then it means that change streams return all changes from the "beginning of time" unless `resumeAfter`/`startAfter`/`startAtOperationTime` is specified, and I failed to find any mention of this behavior.

Reply:
Thanks. So at the end of the day, the data read by `MongoStreamPartitionReader` is determined by the `Offset` passed to `ContinuousStream.planInputPartitions`. There is no concept of reading everything in `ContinuousStream`, and therefore it's fine that we cannot read "everything" via the change stream either.

Reply:
Correct.